Devnet CI

jcompagni10 opened the pull request.

Commit:
add heighliner build and push to CI

Commit:
new TAG env var

Commit:
fix docker push and workspace attach

Commit:
don't tag heighliner image

Commit:
remove extra var

Commit:
Switch to github container registry

Commit:
Sqaushed working devnet setup

dib542 commented on dev.Dockerfile:
```diff
@@ -0,0 +1,18 @@
+
+USER root
+
+CMD ["sh", "./scripts/startup.sh"]
```
we could change this to `leader-startup.sh` and delete `startup.sh` yes?

dib542 commented on scripts/config_setup.sh:
```diff
@@ -0,0 +1,15 @@
+
+echo "Moniker: $NODE_MONIKER"
+
+dasel put bool -f ${HOME}/.duality/config/app.toml -s ".api.enable" "true"
```
I was going to ask where on earth did dasel come from in this Docker. But I realize now its from `ghcr.io/strangelove-ventures/infra-toolkit:v0.0.7` and that was very unclear. could we make a note either here or at import of dasel about what is happening?

dib542 commented on networks/duality-devnet/mnemonics.txt:
```diff
@@ -0,0 +1,4 @@
+reveal gloom baby scan escape shine flock kind marriage sad clinic affair invite dove document youth weekend brave mutual chief sunny return myself sorry
```
since this is needed locally to create the resources I guess this is ok. 

but if we put the docker-compose resource creation step in CI then we could move the mnemonics to the CI environment variables right?

jcompagni10 replied:
I've switched them to environment variables and moved them docker-compose.override.yml for clarity/simplicity. But I'm inclined to keep them in the repo so there is an easy record of what they are. Otherwise, it's kinda PITA to hunt them down later. I suppose there's a risk that someone could find the devnet and do something annoying, but I don't really see a big problem.

dib542 replied:
I like this setup 👍 its both explicit about the mnemonics used but its also very clear that the mnemonics are optional and we could override them in any build by passing other environment variables.

dib542 reviewed (changes requested):
I am/was a bit confused on a few points

- is the docker-compose section necessary? from what I think I understand, the docker-compose and override allows us to generate a cloudformation template with a surprisingly small amount of docker-compose settings and a lot of magic. But we do this from a dev's machine and not from CI. So a dev can stand up a group of resources and the CI then updates them. then that leads into my next point:

- why separate out the `startup.sh` script for only the docker-compose scripts? I don't like that we're duplicating the logic across both paths there. If its important for the default CMD to accept both pathways then we could turn `startup.sh` into a very short script that imports either the `leader-startup.sh` or `follow-startup.sh` depending on the given environment variables. It seems that the two services keep referring to the separate scripts for startup as defined in `docker-compose.yml` after being updated by CI, but that was a bit unclear.

The unclear part is that it is the docker-compose settings being used as the ECS service settings. A bit of documentation and links about how this work would be very beneficial. It took me a good 2 hours of reading this PR to realize they were even related. That seems silly in retrospect but I thought the docker-compose was standing up temporary AWS resources that would disappear the moment that the docker-compose process was ended. 

I would like:
- a few more explanations in the code about "what" we're trying to achieve at steps
- I think there's an issue with the build tagging, we should always tag the commit hashes
- I think the p2p port is not working for its intended purpose and we should consider removing it from devnets
- I'd like to resolve (can we?) that there are two almost identical startup paths. probably the best way forward is to gut `startup.sh` and replace it with sub-startup script imports, so that anyone still referencing startup.sh isn't left behind.

jcompagni10 commented on .circleci/config.yml:
```diff
@@ -58,9 +58,9 @@ jobs:
-          name: Push to ECR
+          name: Push to GHCR
           command: |
             TAG="${CIRCLE_TAG:-${CIRCLE_SHA1}}"
             docker tag duality:local ghcr.io/duality-labs/duality:${TAG}
```
> * is the docker-compose section necessary? from what I think I understand, the docker-compose and override allows us to generate a cloudformation template with a surprisingly small amount of docker-compose settings and a lot of magic. But we do this from a dev's machine and not from CI. So a dev can stand up a group of resources and the CI then updates them. 

This is what I was using to standup the devnet using `docker-compse up` (with ecs context)  since the CI process is not able to do this by itself. Technically it could, but simply updating the image tag on a running devnet is a much smaller task than the full cloud formation setup. Arguably this doesn't have to exist in the codebase, but it seems like a convenient place to keep it. It also makes it very easy to spin up a local devnet.

jcompagni10 commented:
> is the docker-compose section necessary? from what I think I understand, the docker-compose and override allows us to generate a cloudformation template with a surprisingly small amount of docker-compose settings and a lot of magic. But we do this from a dev's machine and not from CI. So a dev can stand up a group of resources and the CI then updates them.
This is what I was using to standup the devnet using docker-compse up (with ecs context) since the CI process is not able to do this by itself. Technically it could, but simply updating the image tag on a running devnet is a much smaller task than the full cloud formation setup. Arguably this doesn't have to exist in the codebase, but it seems like a convenient place to keep it. It also makes it very easy to spin up a local devnet.

dib542 commented on scripts/startup.sh:
```diff
@@ -21,227 +23,48 @@ dualityd init --chain-id $CHAIN_ID duality
+# Start or join a chain
+# note: "new" should not be used for mainnet chains
+# for mainnets a custom genesis file should be curated outside of these scripts
+if [ $STARTUP_MODE == "new" ]
+then
```
tiny formatting difference, all the other `if` statements nearby use `; then` statement on the same line.

dib542 reviewed (changes requested):
It looks good 👍 

there's a small error in `scripts/config_setup.sh` though that should be fixed. 

with that and another comment around the required update of load balancer rules in `networks/duality-devnet/docker-compose.override.yml` then I think its good to go and I'll stamp. 

the inclusion of the indexer could be here but I don't mind if its a separate PR.

teddyknox commented on .circleci/config.yml:
```diff
@@ -148,6 +150,73 @@ jobs:
+      - aws-cli/install
+
+      # Build our network compatible docker image
+      - aws-ecr/build-and-push-image:
```
Does this step rebuild? If so, isn't that redundant with the build job above?

teddyknox commented on .circleci/config.yml:
```diff
@@ -148,6 +150,73 @@ jobs:
             ghr -t "${GITHUB_TOKEN}" -u "${CIRCLE_PROJECT_USERNAME}" -r "${CIRCLE_PROJECT_REPONAME}" -c "${CIRCLE_SHA1}" -delete ${CIRCLE_TAG} ./tmp/artifacts/release
+
+
+  deploy-network:
+    machine:
+      image: ubuntu-2004:2023.04.2
+    working_directory: ~/repo
+    parameters:
+      network:
+        type: string
```
Could we rename this to chainid?

teddyknox commented on Dockerfile:
What are we using this dockerfile for now that we have standardized on the heighliner build?

teddyknox commented on heighliner.Dockerfile:
```diff
@@ -0,0 +1,31 @@
+COPY scripts scripts
+COPY networks networks
+
+CMD ["sh", "./scripts/startup.sh"]
```
Can we move away from this startup-script based setup and towards a volume mounted setup?

dib542 committed:
fix create-gauge CLI number of expected args (#431)

Also make Long description consistent with other descriptions

Commit:
change api listen address to 0.0.0.0

This fixes an issues where otherwise the api is not accesible outside of a
docker container. Likely an issue with some kind of blocking

Commit:
add faucet to docker-compose

Commit:
add pricebot and faucet to compose

Commit:
switch to amd64 build for dockerfile

Commit:
fix unstake cli argument number

previously the dcli builder was defaulting the required number of args to 1.
This cmd should take between 0 & ∞ args

Commit:
fix unstake msg owner addr parsing

Commit:
remove non-deterministic iteration of hash-map from getStakeRefKeys

In specific circumstances (ie. insufficient gas) the non-determinism introduced
by iterating over a hashmap can cause a consensus failure

see: https://go.dev/blog/maps#iteration-order

Commit:
FIX: trader account setup in leader-startup
